### PR TITLE
Adding ZoneColor to set a dynamic color override to an existing zone.

### PR DESCRIPTION
**Pull request by benvanik**

### CODE REVIEW

File: common/TracyQueue.hpp [hunk 1]
```diff
@@ -124,6 +125,10 @@ struct QueueZoneValidation
     uint32_t id;
 };
 
+struct QueueZoneColor {
+  uint32_t value;
```
Comment (wolfpld) [inline common/TracyQueue.hpp:129]: This should be 3 bytes (alpha is ignored).

File: server/TracyEvent.hpp [hunk 2]
```diff
@@ -207,6 +207,7 @@ struct ZoneExtra
     Int24 callstack;
     StringIdx text;
     StringIdx name;
+    uint32_t color;
```
Comment (wolfpld) [inline server/TracyEvent.hpp:210]: Same here, use `Int24`.

File: server/TracyView.cpp [hunk 3]
```diff
@@ -16561,8 +16561,12 @@ uint32_t View::GetRawZoneColor( const ZoneEvent& ev, uint64_t thread, int depth
...
     const auto sl = ev.SrcLoc();
     const auto& srcloc = m_worker.GetSourceLocation( sl );
-    const auto color = srcloc.color;
-    if( color != 0 && !m_vd.forceColors ) return color | 0xFF000000;
+    if( !m_vd.forceColors ) {
+      const auto custom_color = m_worker.GetZoneExtra( ev ).color;
+      if( custom_color != 0 ) return custom_color | 0xFF000000;
+      const auto color = srcloc.color;
+      if( color != 0 ) return color | 0xFF000000;
+    }
```
Comment (wolfpld) [inline server/TracyView.cpp:16569]: Formatting.

File: server/TracyWorker.cpp [hunk 4]
```diff
@@ -4738,6 +4741,22 @@ void Worker::ProcessZoneName()
     extra.name = StringIdx( GetSingleStringIdx() );
 }
 
+void Worker::ProcessZoneColor( const QueueZoneColor& ev )
+{
+    auto td = RetrieveThread( m_threadCtx );
+    if( !td || td->stack.empty() || td->nextZoneId != td->zoneIdStack.back() )
+    {
+        ZoneNameFailure( m_threadCtx );
```
Comment (wolfpld) [inline server/TracyWorker.cpp:4749]: This is wrong.

File: server/TracyView.cpp (hunk 3 above)
Comment (benvanik) [reply server/TracyView.cpp:16569]: sorry not seeing it - what would you change? (I tried to match the existing formatting)

File: server/TracyView.cpp [hunk 5]
```diff
@@ -16561,8 +16561,12 @@ uint32_t View::GetRawZoneColor( const ZoneEvent& ev, uint64_t thread, int depth
 {
     const auto sl = ev.SrcLoc();
     const auto& srcloc = m_worker.GetSourceLocation( sl );
-    const auto color = srcloc.color;
-    if( color != 0 && !m_vd.forceColors ) return color | 0xFF000000;
+    if( !m_vd.forceColors ) {
+      const auto custom_color = m_worker.GetZoneExtra( ev ).color;
```
Comment (wolfpld) [inline server/TracyView.cpp:16565]: `HasZoneExtra( ev ) && GetZoneExtra( ev )`

File: server/TracyView.cpp (hunk 3 above)
Comment (wolfpld) [reply server/TracyView.cpp:16569]: Move the first { brace to a new line and use 4 spaces to indent instead of 2.

Comment (benvanik) [reply server/TracyView.cpp:16569]: ahh nevermind 4 space indent - got it!

File: common/TracyQueue.hpp (hunk 1 above)
Comment (benvanik) [reply common/TracyQueue.hpp]: ah - I didn't see Int24 being used with color [here](https://github.com/wolfpld/tracy/blob/4eb489508def47ca4c64af099050a3607245e0c6/server%2FTracyEvent.hpp#L167) so thought there may have been a reason - changed!

Comment (wolfpld) [reply common/TracyQueue.hpp]: Source locations are low volume, so it doesn't matter that much if everything is optimally packed. Zone colors can be sent for each zone.

Comment (benvanik) [reply common/TracyQueue.hpp]: It looks like Int24 isn't accessible from common/ (it's defined in server/TracyEvent.hpp) - should I just use uint8_t[3] and cast it around?

Comment (benvanik) [reply common/TracyQueue.hpp]: nevermind, I see SourceLocation has a split r g b - I'll do the same.

**Pull request merged**